sdk_contrib/koa/lib/koa_mw.js

```diff
@@ -0,0 +1,113 @@
+ * @module koa_mw
+ */
+
+const AWSXRay = require("aws-xray-sdk-core");
+
+const mwUtils = AWSXRay.middleware;
+const IncomingRequestData = mwUtils.IncomingRequestData;
+const Segment = AWSXRay.Segment;
+
+const koaMW = {
+  /**
+   * Use 'app.use(AWSXRayKoa.openSegment(<name>))' before defining your routes.
+   * Use AWSXRay.getSegment() to access the current sub/segment.
+   * Otherwise, for manual mode, this appends the Segment object to the request object as ctx.segment.
+   * @param {string} defaultName - The default name for the segment.
+   * @alias module:koa_mw.openSegment
+   * @returns {function}
+   */
+
+  openSegment: function openSegment(defaultName) {
+    if (!defaultName || typeof defaultName !== "string")
+      throw new Error(
+        "Default segment name was not supplied.  Please provide a string."
+      );
+
+    mwUtils.setDefaultName(defaultName);
+
+    return async (ctx, next) => {
+      const amznTraceHeader = mwUtils.processHeaders(ctx);
+      const name = mwUtils.resolveName(ctx.host);
+      const segment = new Segment(
+        name,
+        amznTraceHeader.Root || amznTraceHeader.root,
+        amznTraceHeader.Parent || amznTraceHeader.parent
+      );
+
+      mwUtils.resolveSampling(amznTraceHeader, segment, ctx);
+      segment.addIncomingRequestData(new IncomingRequestData(ctx.req));
+
+      exports._logToXRay("Starting koa segment", ctx, segment);
+
+      if (AWSXRay.isAutomaticMode()) {
+        const ns = AWSXRay.getNamespace();
+        ns.bindEmitter(ctx.req);
+        ns.bindEmitter(ctx.res);
+
+        return ns.runAndReturn(async function () {
+          let error;
+          AWSXRay.setSegment(segment);
+          ctx.segment = segment;
+          try {
+            if (next) {
+              await next();
+            }
+          } catch (err) {
+            error = err;
+          }
+          exports._processResponse(ctx, segment, error);
+        });
+      } else {
+        let error;
+        ctx.segment = segment;
+        try {
+          if (next) {
+            await next();
+          }
+        } catch (err) {
+          error = err;
+        }
+        exports._processResponse(ctx, segment, error);
+      }
+    };
+  },
+};
+
+exports._processResponse = (ctx, segment, err) => {
+  if (ctx.status >= 400) {
+    if (ctx.status === 429) {
+      segment.addThrottleFlag();
+    }
+    segment[AWSXRay.utils.getCauseTypeFromHttpStatus(ctx.status)] = true;
+  }
+
+  if (segment.http && ctx.res) {
+    segment.http.close(ctx.res);
+  }
+  segment.close(err);
+  exports._logToXRay(
+    err ? "Closed koa segment with error" : "Closed koa segment successfully",
+    ctx,
+    segment
+  );
+  if (err) {
+    throw err;
+  }
+};
+
+exports._logToXRay = (message, ctx, segment) => {
+  AWSXRay.getLogger().debug(
+    `${message}: { url: ${ctx.url}, name: ${segment.name}, trace_id: ${
+      segment.trace_id
+    }, id: ${segment.id}, sampled: ${!segment.notTraced} }`
+  );
+};
+
```

Comment:
@willarmiros I have made this change, but there is a problem: 
The projects in sdk_contrib are not part of the Lerna project so they do not get the changes from the main library sym-linked automatically the way a project in the "packages" folder would.  This log functionality from core is not yet in a released version, so my tests fail.  I had to make a file-based dev-dependency in the package.json to circumvent this issue: 
`"aws-xray-sdk-core": "file:../../packages/core", `

This works fine, but if somebody pulls in a peer-dependency of 3.1.0, it will not work as the middlewareLog function is not present in 3.1.0.  So the best thing I can do is assume the next version will be 3.2.0, and I have set that as the peer dependency.

Reply:
hmm that's interesting, I included the `sdk_contrib` in lerna [here](https://github.com/aws/aws-xray-sdk-node/blob/master/lerna.json#L6). I tested it with the corrected 3.1.0 versions and it seemed to work locally.

Proposed change: https://github.com/unbill/aws-xray-sdk-node/pull/1

Reply:
Oh shoot my bad...I didn't even try doing a Lerna build...I just assumed the sdk_contrib was outside the scope of Lerna...so no issues.

Reply:
Yep after lerna bootstrap --hoist, the tests run fine...perfect.